### PR TITLE
Ensure variants returned in #variant_relation are DISTINCT

### LINKED ISSUE #4413: Shopfronts Not Loading all products
## Description

With the new release Tamar Valley Food Hub shopfront is not loading all products.

The shopfront scrolls to a certain point and then stops. The spinner spins for a while, then it times out without all the products there.

Run this with DevTools and you'll see and interesting and indicative error:

Screenshot from 2019-10-29 15-42-27

I suggest you just load this shop yourself to see it:
https://openfoodnetwork.org.uk/tamar-valley-food-hubs/shop

## Expected Behavior

We see an endless scroll of products loading without error

## Actual Behaviour

After 3 pages of products the shopfront fails to load further

## Steps to Reproduce

1. Go to https://openfoodnetwork.org.uk/tamar-valley-food-hubs/shop
2. Scroll Down
3. They should have 500+ products in their shop
4.

## Severity

S1

## Your Environment

* Version used: Mercao Rapido
* Browser name and version: Chrome and Firefox
* Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #4413

Fixes an issue with shopfront pagination where duplicates were sometimes returned in the results. This was present in OCs where; 1) multiple products from the same supplier in the OC had the exact same name and 2) the hub was using the feature for reordering products by a custom taxon list.

The issue was discovered on UK production. It probably had minimal impact beyond a single OC where both of these conditions were met.

#### What should we test?

Deploying `v2.6.0` to UK Staging will show the bug, when viewing the shopfront for Tamar Valley Food Hub and scrolling down far enough. The javascript console will show an error and the loading spinner will be stuck, when reaching the `goats milk` product.

Deploying this PR to UK Staging will fix the issue. Scrolling to the bottom of the full product list is possible. Both versions of the duplicate `goats milk` product will be displayed correctly.

#### Release notes

Fixed an issue with duplicate product names affecting shopfront pagination

Changelog Category: Fixed

### CODE REVIEW

**Comment by luisramos0** [comment]
nice!!!!

we need some for of testing to validate the issue and the fix.

looking at the code I wonder/suspect Api::ProductSerializer#master is not needed and quite a few lines of code can be deleted?

File: app/services/order_cycle_distributed_products.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ def initialize(distributor, order_cycle, customer)
   end
 
   def products_relation
-    Spree::Product.where(id: stocked_products)
+    Spree::Product.where(id: stocked_products).group("spree_products.id")
```
Comment (mkllnk) [inline app/services/order_cycle_distributed_products.rb]: Is this is an unusual use of `group`? Why didn't you go for the same approach as before? `select("spree_products.*")`

Comment (Matt-Yorkley) [reply app/services/order_cycle_distributed_products.rb]: I wanted to make sure there were no duplicates in the results, and initially tried using DISTINCT in the query here, but I ran into an obscure Rails/Postgres issue when the taxon ordering feature is applied. It caused a lot of fatal errors and took a while to figure out. I settled on using group by id as a workaround.

Further reading: 

https://stackoverflow.com/questions/12693089/pgerror-select-distinct-order-by-expressions-must-appear-in-select-list/18367248

https://github.com/rails/rails/issues/520

**Comment by Matt-Yorkley** [comment]
I couldn't make a working spec for this initially, but I'll try again.

**Comment by Matt-Yorkley** [comment]
Nope, I can't replicate this bug in a test...

**Comment by Matt-Yorkley** [comment]
Okay, I've got it. Ready for review!

**Comment by luisramos0** [comment]
@Matt-Yorkley the new specs are green in master... I'll try to understand what's missing.

**Comment by kristinalim** [comment]
The specs are also green for me in master...

The second commit touching the spec (67a5a1cdc2e809fce34ff097bbdc686f8606787d) causes a failure for me, but unrelated:

```
  1) Api::OrderCyclesController with custom taxon ordering applied and duplicate product names in the order cycle displays products in correct order across multiple pages
     Failure/Error: expect(product_ids).to eq [product7.id, product8.id, product2.id]

       expected: [249, 250, 244]
            got: [250, 249, 244]

       (compared using ==)
     # ./spec/controllers/api/order_cycles_controller_spec.rb:212:in `block (3 levels) in <module:Api>'
```

**Comment by luisramos0** [comment]
all right, took me a while to get this right!

First, this will fix our problem!

I have this commit that changes the spec:
https://github.com/Matt-Yorkley/openfoodnetwork/pull/8/commits/d1a01b12b199ad0f2411fa43a37e3f906bd2599c

see the commit message

I also wrote a feature spec to replicate this but I think this new controller spec is enough 💪

File: app/services/order_cycle_distributed_products.rb (hunk 1 above)
Comment (mkllnk) [reply app/services/order_cycle_distributed_products.rb]: Ah, okay, I ran into that issue before. It's frustrating.

**Comment by filipefurtad0** [comment]
Hi @Matt-Yorkley,

Followed your testing instructions: worked just as described :tada :-)
-> moving to Ready to Go

For pics and testing notes please see:
https://docs.google.com/document/d/1DVuxciWJL4DN0_XpmZ5RigS1tzsTyZBqyOOj_s2fxDA/

Cheers!

**Comment by luisramos0** [comment]
great Filipe!

**Pull request merged**